wagtail/images/blocks.py

```diff
@@ -51,3 +57,156 @@ def htmldiff(self):
+
+                break  # no need for further execution after bulk_to_python has been called.
+
+            else:
+                struct_values = super().bulk_to_python(values)
+                break  # no need for further execution after bulk_to_python has been called.
+
+        return [
+            self._struct_value_to_image(struct_value) for struct_value in struct_values
+        ]
+
+    def value_from_datadict(self, data, files, prefix):
+        struct_value = super().value_from_datadict(data, files, prefix)
+        return self._struct_value_to_image(struct_value)
+
+    def clean(self, value):
+        if value is None:
+            raise StructBlockValidationError(
+                block_errors={
+                    "image": ValidationError(
+                        _("Expected an image instance, got nothing")
+                    )
+                }
+            )
+
+        if not isinstance(value, AbstractImage):
+            raise StructBlockValidationError(
+                block_errors={
+                    "image": ValidationError(
+                        _("Expected an image instance, got %(value)s")
+                        % {"value": value}
+                    )
+                }
+            )
+
+        if not value.contextual_alt_text and not value.decorative:
+            raise StructBlockValidationError(
+                block_errors={
+                    "alt_text": ValidationError(
+                        _(
+                            "Please add some alt text for your image or mark it as decorative"
+                        )
+                    )
+                }
+            )
+
+        return value
+
+    def normalize(self, value):
+        if value is None or isinstance(value, AbstractImage):
+            return value
+        else:
+            struct_value = super().normalize(value)
+            return self._struct_value_to_image(struct_value)
+
+    def get_form_context(self, value, prefix="", errors=None):
+        dict_value = {
+            "image": value,
+            "alt_text": value and value.contextual_alt_text,
+            "decorative": value and value.decorative,
+        }
+        context = super().get_form_context(dict_value, prefix=prefix, errors=errors)
+        context["suggested_alt_text"] = value
+        return context
+
+    def get_form_state(self, value):
+        return {
+            "image": self.child_blocks["image"].get_form_state(value),
+            "alt_text": value and value.contextual_alt_text,
+            "decorative": value and value.decorative,
+        }
+
+    def get_prep_value(self, value):
+        return {
+            "image": self.child_blocks["image"].get_prep_value(value),
+            "alt_text": value and value.contextual_alt_text,
+            "decorative": value and value.decorative,
+        }
+
+    def extract_references(self, value):
+        return self.child_blocks["image"].extract_references(value)
+
```

Comment:
We’re missing an implementation of `get_api_representation`. In my testing, seems like the following would work well. It’s not backwards-compatible, but we couldn’t return the contextual alt text in a backwards-compatible way anyway:

```python
    def get_api_representation(self, value, context=None):
        return self.get_prep_value(value)
```

Reply:
Implemented in https://github.com/wagtail/wagtail/pull/12416/commits/f09fb3d22aa2fa54264ec6e148322b0b92377fe8